jmix-security/security-flowui/src/main/java/io/jmix/securityflowui/authentication/LoginViewSupport.java

```diff
@@ -269,12 +281,60 @@ protected Location getRedirectLocation(VaadinServletRequest request, VaadinServl
 
         SavedRequest savedRequest = requestCache.getRequest(httpServletRequest, response);
         if (savedRequest != null) {
-            return new Location(savedRequest.getRedirectUrl());
+            if (savedRequest instanceof DefaultSavedRequest defaultSavedRequest) {
+                QueryParameters queryParameters = QueryParameters.fromString(defaultSavedRequest.getQueryString());
+                return new Location(defaultSavedRequest.getServletPath(), queryParameters);
+            } else {
+                return new Location(savedRequest.getRedirectUrl());
+            }
         }
 
         return null;
     }
 
+    protected boolean isRedirectToInitialView(Location redirectLocation) {
+        if (StringUtils.isNotEmpty(redirectLocation.getQueryParameters().getQueryString())) {
+            return false;
+        }
+        RouteConfiguration routeConfiguration = RouteConfiguration.forSessionScope();
+        return routeConfiguration.getRoute(redirectLocation.getPathWithQueryParameters())
+                .map(StandardMainView.class::isAssignableFrom)
+                .orElse(false);
+    }
+
+    protected void navigateToInitialView() {
+        String defaultViewId = uiProperties.getDefaultViewId();
+        if (StringUtils.isEmpty(defaultViewId)) {
```

Comment:
`com.google.common.base.Strings` the preferable choice for String utility methods.